compiler/src/iree/compiler/Dialect/Util/Transforms/CombineInitializers.cpp

```diff
@@ -4,77 +4,308 @@
+        if (auto symRef = dyn_cast<SymbolRefAttr>(callee)) {
+          if (auto calleeOp = symbolTable.lookup<FunctionOpInterface>(
+                  symRef.getRootReference())) {
+            // Avoid infinite recursion by tracking visited functions.
+            if (visitedFuncs.insert(calleeOp.getOperation()).second) {
+              worklist.push_back(&calleeOp.getFunctionBody());
+            }
+          }
+        }
+      }
+    });
   }
 
+  return globals.takeVector();
+}
+
+struct CombineInitializersPass
+    : public impl::CombineInitializersPassBase<CombineInitializersPass> {
   void runOnOperation() override {
     mlir::ModuleOp moduleOp = getOperation();
 
-    // Gather all of the initializers in the module.
-    // Build a fused loc from all initializers we are combining.
-    SmallVector<IREE::Util::InitializerOp> initializerOps;
+    // Gather all initialization points in module order.
+    // This includes both util.initializer ops and util.global ops with initial
+    // values.
+    SmallVector<Operation *> initPoints;
     SmallVector<Location> locs;
-    for (auto initializerOp : moduleOp.getOps<IREE::Util::InitializerOp>()) {
-      initializerOps.push_back(initializerOp);
-      locs.push_back(initializerOp.getLoc());
+    bool hasInitializers = false;
+    for (auto &op : moduleOp.getOps()) {
+      if (isa<IREE::Util::InitializerOpInterface>(&op)) {
+        initPoints.push_back(&op);
+        locs.push_back(op.getLoc());
+        hasInitializers = true;
+      } else if (auto globalOp = dyn_cast<IREE::Util::GlobalOpInterface>(&op)) {
+        auto initialValue = globalOp.getGlobalInitialValue();
+        if (initialValue && !isa<IREE::Util::UninitializedAttr>(initialValue)) {
+          initPoints.push_back(&op);
+          locs.push_back(op.getLoc());
+        }
+      }
     }
-    if (initializerOps.size() <= 1)
+
+    // We only proceed if we have initializers - if we only have globals with
+    // initial values the module is already in a stable state.
+    if (!hasInitializers || initPoints.size() <= 1) {
       return;
-    auto fusedLoc = FusedLoc::get(&getContext(), locs);
-
-    // Make the new initializer op in the same location as the first initializer
-    // we are combining - this ensures that module initialization order is
-    // preserved.
-    OpBuilder builder(initializerOps.front());
-    auto newOp = IREE::Util::InitializerOp::create(builder, fusedLoc);
-    builder.setInsertionPointToStart(newOp.addEntryBlock());
+    }
+
+    // Simple symbol table for lookups.
+    // We are otherwise performing local operations and do not need a full
+    // GlobalTable analysis.
+    SymbolTable symbolTable(moduleOp);
+
+    // Process initialization points, potentially creating multiple combined
+    // initializers if we encounter non-materializable constants that we can't
+    // place inside our combined initializer. Initialization order of the
+    // program will be preserved.
```

Comment:
Question outside this PR's scope: is it the place where we break big initializers into several initializers (or say not combining the initializers), if it exceeds memory constraints? I don't know how to query such information, though. It is not described in an executable target, which seems reasonable.